README
Test::Spec
==========

This is a declarative specification‐style testing system for behavior‐
driven development (BDD) in Perl. The tests (a.k.a. examples) are named
with strings instead of subroutine names, so your fingers will suffer
less fatigue from underscore−itis, with the side benefit that the test
reports are more legible.

This module is inspired by and borrows heavily from RSpec
(http://rspec.info/documentation/), a BDD tool for the Ruby programming
language.

See `perldoc Test::Spec` for syntax examples and usage information.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

 * constant
 * List::Util
 * Package::Stash (>= 0.23)
 * Scalar::Util (XS version)
 * TAP::Parser
 * Test::Deep (>= 0.103)
 * Test::More
 * Test::Trap
 * Tie::IxHash

AUTHOR

Philip Garrett <philip.garrett@icainformatics.com>

SOURCE CODE

The source code for Test::Spec lives at github:
  https://github.com/kingpong/perl-Test-Spec

COPYRIGHT AND LICENSE

Copyright (C) 2011 by Informatics Corporation of America.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Test::Spec.

0.54 Tue Nov 21 12:46:00 MSK 2017
  - `around` method added to allow localized variables and other local setup.
    Call `yield` method to run examples by `around` CODE.
    Contributed by @akzhan

  - Replace some `use base` with `use parent` and other minor updates.
    Contributed by @akzhan

  - Now tested under Perl 5.22-5.26 too
    Contributed by @akzhan

  - predictable_destroy spec fixed under Perl 5.8
    Contributed by @akzhan

0.53 Wed Aug 16 09:49:00 BST 2017
  - POD links are clickable
    Contributed by @manwar (issue #42)

0.52 Tue Aug 15 10:20:00 BST 2017
  - Pass example instance to each test
    Contributed by @jamhed (issue #30, #41)
  - Fix runtests to run all examples that match any of the patterns
    Contributed by @mpw96 (issue #37, #38)
  - Ensure META.json includes requirements
    Reported by @joelrebel (issue #39)

0.51 Mon Jul 27 20:05:00 BST 2015
  - Fixed top level describe blocks running out of order
    Contributed by @mla (issue #28)

0.50 Sun Apr 19 18:13:00 BST 2015
  - Removed several memory leaks
    Reported by @JRaspass (issue #24)
  - any_number() no longer fails if it is never called
    Contributed by @ptolemarch and @cavemanpi (issue #25)

0.49 Sun Jan 18 12:56:00 EST 2015
  - Added with_deep
    Contributed by Andy Jones (issue #21)
  - Documented memory leak in stub()
    Reported by Victor Efimov (issue #14)
  - Added Travis CI info
    Contributed by Andy Jones (issue #23)
  - Fixed mock() example documentation
    Reported by Victor Efimov (issue #15)
    Contributed by Andy Jones (issue #22)

0.48 Tue Jan 06 22:52:00 EST 2015
  - Testing fixes for Alpha branch of Test::More
    Contributed by Chad Granum (issue #18)
  - Metadata fixes
    Contributed by Graham Knop (issue #19)

0.47 Fri Oct 11 01:22:00 EDT 2012
  - added repository metadata for CPAN
    Contributed by David Steinbrunner (issue #13)

0.46 Tue Oct  2 13:23:00 EDT 2012
  - with() enables argument matching on mocked methods
  - raises() makes mocked methods raise exceptions
    Contributed by Kjell-Magne Øierud (issue #12)

0.45 Mon May  7 10:08:13 EDT 2012
  - Add support for TAP version 13.
    Contributed by Michael G. Schwern (issue #11)

0.44 Mon Apr 30 11:04:00 CST 2012
  - Allow shared_examples_for to be defined in any context.

0.43 Sat Apr 14 16:22:00 EST 2012
  - Fixed runtests() to honor its contract to run only the examples specified
    in its @patterns parameter or SPEC environment variable.

0.42 Mon Mar 05 21:18:00 CST 2012
  - Added context() and xcontext() aliases for describe/xdescribe
    (reported by intrigeri)

0.41 Sat Mar 03 19:04:00 EST 2012
  - Added license info to Makefile.PL (RT #75400)
  - Fixed test suite problems on Windows

0.40 Mon Jan 30 18:38:00 EST 2012
  - Fixed problem that caused Test::Spec usage errors (e.g. 'describe "foo";'
    without a subroutine argument) to be reported from inside the library,
    instead of the caller's perspective where the actual error is.

0.39 Wed Aug 31 00:52:00 EST 2011
  - Added xit/xthey/xdescribe to mark TODO tests, inspired by the
    Jasmine JavaScript framework.
    Contributed by Marian Schubert (issue #10).

0.38 Sat Jul 09 23:16:00 EST 2011
  - Added share() function to facilitate spec refactoring.

0.37 Thu Jul 07 13:55:00 EST 2011
  - Fixed bug where shared examples defined in one package could not be
    used in another package.

0.36 Tue Jul 05 18:23:00 EST 2011
  - Improved reporting of errors using spec_helper.
  - Minor documentation formatting fixes

0.35 Wed Jun 29 16:52:00 UTC 2011
  - Fixed test suite for Windows environments.

0.34
  - Added spec_helper utility function to load helper scripts
    relative to the spec.

0.33 Mon Jun 13 15:03:03 UTC 2011
  - Added shared_examples_for/it_should_behave_like to allow
    factorization of tests.

0.32 Thu Jun  9 16:09:55 UTC 2011
  - Fixed a problem with the tests that occurred only when
    Package::Stash::XS was not installed (issue #8).

0.31
  - "use Test::Spec" now enables strict and warnings in the calling
    scope to lessen boilerplate code in specs.
    Contributed by Marian Schubert (issues #2, #3, #4, #7)
  - 6x speed improvement by removing Moose (test suite went from 8.33s
    to 1.25s).
  - Diagnostics improvements (issue #5)

0.30  Fri Jun  3 04:55:54 2011
  - Examples that die will now display the error message even when run
    through a harness like prove(1).

0.29  Thu May 19 18:49:00 2011
        - Quell annoying Test::Deep::isa() warnings.

0.28  Thu May 19 11:15:58 2011
        - Extracted from ICA::Test::Spec



t/dying_spec.pl
#!/usr/bin/env perl
#
# dying_test.pl
#
# Expected to fail. It should output TAP in such a way that prove(1)
# will display the exception message.
#
########################################################################
#

use Test::Spec;
use FindBin qw($Bin);
BEGIN { require "$Bin/test_helper.pl" };

describe "Test::Spec" => sub {
  it "should trap die message" => sub {
    die "this should be displayed";
  };
  it "should continue testing" => sub {
    ok(1);
  };
};

runtests unless caller;


t/helper_test.pl
#
# just increment the value of $foo in the current package.
#
no strict;
$foo++;


Makefile.PL
use 5.008005;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
my %MM_ARGS = (
    NAME              => 'Test::Spec',
    VERSION_FROM      => 'lib/Test/Spec.pm', # finds $VERSION
    PREREQ_PM         => {
      'Carp'           => 0,
      'Exporter'       => 0,
      'List::Util'     => 0,
      'Package::Stash' => 0.23,
      'Scalar::Util'   => 1.11,
      'Test::Deep'     => 0.103, # earlier versions clash with UNIVERSAL::isa
      'Test::More'     => 0.88,
      'Test::Trap'     => 0,
      'Tie::IxHash'    => 0,
      'constant'       => 0,
    },
    TEST_REQUIRES => {
      'TAP::Parser'        => 0,
      'Devel::GlobalPhase' => 0,
    },
    ABSTRACT_FROM  => 'lib/Test/Spec.pm', # retrieve abstract from module
    AUTHOR         => 'Philip Garrett <philip.garrett@icainformatics.com>',
    LICENSE => 'perl',
    META_MERGE        => {
        'meta-spec' => { version => 2.0 },
        resources => {
            repository => {
                type => 'git',
                web => 'https://github.com/kingpong/perl-Test-Spec',
                url => 'https://github.com/kingpong/perl-Test-Spec',
            },
            bugtracker => {
                web => 'https://github.com/kingpong/perl-Test-Spec/issues',
            },
        },
    },
);

$MM_ARGS{BUILD_REQUIRES} = {%{$MM_ARGS{BUILD_REQUIRES}}, %{delete $MM_ARGS{TEST_REQUIRES}}}
  if $eumm_version < 6.63_03;
$MM_ARGS{PREREQ_PM} = {%{$MM_ARGS{PREREQ_PM}}, %{delete $MM_ARGS{BUILD_REQUIRES}}}
  if $eumm_version < 6.55_01;
WriteMakefile(%MM_ARGS);
